system/jlib/jhash.hpp

```diff
@@ -598,4 +603,82 @@ template <class C> class CMinHashTable
 };
 
 
+/* 
+ * A HT/Cache implementation whose items are only valid for a defined timeout period (default 30 secs)
+ * Example use:
+ *   CTimeLimitedCache<std::string, Owned<IPropertyTree>> myCache(10);
+ *   Owned<IPropertyTree> match;
+ *   verifyex( !myCache.get("tree1", match) );
+ *   myCache.add("tree1", createPTree("tree1"));
+ *   verifyex( nullptr != myCache.query("tree1") );
+ *   MilliSleep(11000); // sleep until "tree1" has expired
+ *   verifyex( nullptr == myCache.query("tree1") );
+ *
+ *   myCache.ensure("tree2", [](std::string key) { return createPTree(key.c_str()); });
+ *
+ * Keywords: cache,time,limit,hashtable,hash
+ */
+
+template <class KEYTYPE, class VALUETYPE>
+class jlib_decl CTimeLimitedCache
+{
+    static constexpr unsigned defaultCacheTimeoutMs = 30000;
+    typedef std::pair<cycle_t, VALUETYPE> CacheElement;
+    cycle_t timeoutPeriodCycles = 0;
+    std::unordered_map<KEYTYPE, CacheElement> ht;
+
+    CacheElement *getMatch(KEYTYPE key)
+    {
+        auto it = ht.find(key);
+        if (it == ht.end())
+            return nullptr;
+        if ((get_cycles_now() - it->second.first) > timeoutPeriodCycles) // NB: rollover is okay
```

Comment:
For some use-cases, may want to update time last seen each time it is looked up?

Reply:
I actually added queryTouch/getTouch methods initially, then though keep it simple, add a future date if needed/semantics became clearer.
could re-add..

Reply:
Have re-added (see 2nd commit)